### PR TITLE
Upgrader and CLI features

### PR DESCRIPTION
**Pull request by dougblackjr**
## Overview

Per discussion, this PR adds two thing:
1. Upgrade Fixes: This adds
- namespacing to EE2 updates so that they can be run in the EE upgrade process via the CLI (see below) or UI.
- Add-on upgrading during the EE core upgrading process via available add-on hooks
- Database backup as part of the EE core upgrade process.

2. CLI Feature. This allows for creation of commands for use in the command line. Documentation available on docs PR [https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/90](https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/90).

In addition, the CLI feature adds:
- CLI upgrading from EE2 to EE5
- ability to move file structure from EE2 to EE5 similar structure, with a good amount of control, via `prepare-upgrade` command.

## Nature of This Change

- [X] 🐛 Fixes a bug
- [X] 🚀 Implements a new feature
- [X] 🛁 Refactors existing code
- [ ] 💅 Fixes coding style
- [ ] ✅ Adds tests
- [ ] 👽 Adds new dependency
- [ ] 🔥 Removes unused files / code
- [ ] 🔒 Improves security 

## Is this backwards compatible?

- [X] Yes
- [ ] No

## Documentation

User Guide Pull Request: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/90

### CODE REVIEW

File: eecli [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env php
+<?php
+
+define('EE_START', microtime(true));
+
+$system_path = realpath(dirname($_SERVER['SCRIPT_FILENAME'])) . '/system';
+
+if (PHP_SAPI !== 'cli') {
+    header('HTTP/1.1 503 Service Unavailable.', true, '503');
+    exit("Invlid request\n");
```
Comment (obfuscode) [inline eecli:10]: Spelling mistake.

Comment (dougblackjr) [reply eecli]: Fixed on latest commit.

File: system/ee/installer/controllers/wizard.php [hunk 2]
```diff
@@ -33,9 +33,11 @@ class Wizard extends CI_Controller {
...
 	public $subtitle          = '';
 
-	private $current_step = 1;
-	private $steps        = 3;
-	private $addon_step   = FALSE;
+	private $current_step 			= 1;
+	private $steps        			= 3;
+	private $addon_step   			= FALSE;
+	private $shouldBackupDatabase	= false;
+	private $shouldUpgradeAddons 	= false;
```
Comment (obfuscode) [inline system/ee/installer/controllers/wizard.php:40]: Constant's case should match here.

File: system/ee/EllisLab/ExpressionEngine/Library/Filesystem/Filesystem.php [hunk 3]
```diff
@@ -667,6 +667,51 @@ public function getUniqueFilename($path)
...
+
+		$contents = $this->read($file);
+
+		if (strpos($search, '/') === 0) {
+
+		    $contents = preg_replace($search, $replace, $contents);
+
+		} else {
+
+		    $contents = str_replace($search, $replace, $contents);
```
Comment (obfuscode) [inline system/ee/EllisLab/ExpressionEngine/Library/Filesystem/Filesystem.php:707]: Mix of tabs and spaces for whitespace.

File: system/ee/EllisLab/ExpressionEngine/Cli/Cli.php [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    /**
+     * check if command is defined among the addons
+     * @param  string $command [defaults to whatever command was called]
+     * @return bool
+     */
+    protected function commandExists($command = null)
+    {
+        $commandToParse = $command ?? $this->commandCalled;
```
Comment (obfuscode) [inline system/ee/EllisLab/ExpressionEngine/Cli/Cli.php:269]: Null-coalescing operator needs to be refactored to PHP 5.6 compatible.

File: system/ee/EllisLab/ExpressionEngine/Cli/Cli.php [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    /**
+     * get command class
+     * @param  string $command [defaults to whatever command was called]
+     * @return mixed
+     */
+    protected function getCommand($command = null)
+    {
+        $commandToParse = $command ?? $this->commandCalled;
```
Comment (obfuscode) [inline system/ee/EllisLab/ExpressionEngine/Cli/Cli.php:286]: Null-coalescing operator needs to be refactored to PHP 5.6 compatible.

File: system/ee/EllisLab/ExpressionEngine/Cli/Cli.php [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+     * @param  string $question
+     * @return mixed
+     */
+    public function ask($question, $default = '')
+    {
+        $this->output->out($question . ' ');
+
+        $result = $this->input->in();
+
+        return $result ?? $default;
```
Comment (obfuscode) [inline system/ee/EllisLab/ExpressionEngine/Cli/Cli.php:227]: Null-coalescing operator needs to be refactored to PHP 5.6 compatible.

File: system/ee/EllisLab/ExpressionEngine/Cli/Commands/CommandPrepareUpgrade.php [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+    public $commandOptions = [
+        'upgrade-ee'                => 'Start the upgrade after moving files',
+        'force-add-on-upgrade'      => 'After upgrading EE, runs addon upgrades',
+        'old-base-path:'            => 'Absolute path of old site',
+        'new-base-path:'            => 'Absolute path of new site',
+        'old-public-path:'          => 'Absolute path of old site public path',
+        'new-public-path:'          => 'Absolute path of new site public path',
+        'no-config-file'            => 'Ignores the config file and doesn\'t check for it',
+        'ee-version'                => 'The current site ',
+        'should-move-system-path'   => 'Whether the upgrade process should move the old theme folder to the new site',
```
Comment (obfuscode) [inline system/ee/EllisLab/ExpressionEngine/Cli/Commands/CommandPrepareUpgrade.php:55]: This says "theme" but should say "system"

File: system/ee/EllisLab/ExpressionEngine/Cli/Commands/CommandUpdate.php [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+	 * @var string
+	 */
+	public $usage = 'php eecli update';
+
+	/**
+	 * options available for use in command
+	 * @var array
+	 */
+	public $commandOptions = [
+		'rollback'				=> 'Rollsback last update',
```
Comment (obfuscode) [inline system/ee/EllisLab/ExpressionEngine/Cli/Commands/CommandUpdate.php:49]: "Rollsback" has extra S or should be two words

File: system/ee/installer/controllers/wizard.php (hunk 2 above)
Comment (dougblackjr) [reply system/ee/installer/controllers/wizard.php:40]: @obfuscode With the move to PSR-2, are we going lowercase or uppercase?

File: system/ee/EllisLab/ExpressionEngine/Library/Filesystem/Filesystem.php (hunk 3 above)
Comment (dougblackjr) [reply system/ee/EllisLab/ExpressionEngine/Library/Filesystem/Filesystem.php]: Fixed in latest commit.

File: system/ee/EllisLab/ExpressionEngine/Cli/Cli.php (hunk 5 above)
Comment (dougblackjr) [reply system/ee/EllisLab/ExpressionEngine/Cli/Cli.php]: Fixed in latest commit.

File: system/ee/EllisLab/ExpressionEngine/Cli/Cli.php (hunk 6 above)
Comment (dougblackjr) [reply system/ee/EllisLab/ExpressionEngine/Cli/Cli.php]: Fixed in latest commit.

File: system/ee/EllisLab/ExpressionEngine/Cli/Commands/CommandPrepareUpgrade.php (hunk 7 above)
Comment (dougblackjr) [reply system/ee/EllisLab/ExpressionEngine/Cli/Commands/CommandPrepareUpgrade.php:55]: Fixed in latest commit.

File: system/ee/installer/controllers/wizard.php (hunk 2 above)
Comment (obfuscode) [reply system/ee/installer/controllers/wizard.php:40]: I'd say user uppercase for now as we're not refactoring existing code and mixing the two is less desirable.

File: system/ee/EllisLab/ExpressionEngine/Cli/Cli.php (hunk 4 above)
Comment (dougblackjr) [reply system/ee/EllisLab/ExpressionEngine/Cli/Cli.php]: Fixed on latest commits.

File: system/ee/installer/controllers/wizard.php (hunk 2 above)
Comment (dougblackjr) [reply system/ee/installer/controllers/wizard.php]: @obfuscode fixed on latest commit.

**Review by obfuscode — APPROVED** [review]

**Pull request merged**